once/_iterator_wrappers.py

```diff
@@ -0,0 +1,186 @@
+                    # the iterator. We just store the value in self.results so that
+                    # we can later yield it outside of the lock.
+                    assert self.generator is not None
+                    # TODO(matt): Is the fact that we have to suppress typing here a bug?
+                    self.generating = self.generator.asend(send)  # type: ignore
+                    generating = self.generating
+                elif i == len(self.results) and self.finished:
+                    # All done.
+                    return
+                else:
+                    # We already have the correct result, so we grab it here to
+                    # yield it outside the lock.
+                    next_val = self.results[i]
+
+            if generating:
+                try:
+                    next_val = await generating
+                except StopAsyncIteration:
+                    async with self.lock:
+                        self.generator = None  # Allow this to be GCed.
+                        self.finished = True
+                        self.generating = None
+                        generating = None
+                        return
+                async with self.lock:
+                    self.results.append(next_val)
+                    generating = None
+                    self.generating = None
+
+            send = yield next_val
+            i += 1
+
+
+class GeneratorWrapper:
+    """Wrapper around an sync generator which only runs once.
+
+    Subsequent calls will return results from the first call, which is
+    evaluated lazily.
+    """
+
+    def __init__(self, func, *args, **kwargs) -> None:
+        self.generator: collections.abc.Generator | None = func(*args, **kwargs)
+        self.finished = False
+        self.results: list = []
+        self.generating = False
+        self.lock = threading.Lock()
+        self.next_send = None
+
+    def _yield_results(self) -> collections.abc.Generator:
+        i = 0
+        # Fast path for subsequent calls will not require a lock
+        while True:
+            # If we on before the penultimate entry, we can return now. When yielding the last
+            # element of results, we need to be recording next_send, so that needs the lock.
+            if i < len(self.results) - 1:
+                yield self.results[i]
+                i += 1
+                continue
+            # Because we don't hold a lock here, we can't make this assumption
+            # i == len(self.results) - 1 or i == len(self.results)
+            # because the iterator could have moved in the interim. However, it will no longer
+            # move once self.finished.
+            if self.finished:
+                if i < len(self.results):
+                    yield self.results[i]
+                    i += 1
+                    continue
+                if i == len(self.results):
+                    return
```

Comment:
IMO this code is way too fancy trying to avoid grabbing the lock in the happy path. The lock is held super briefly since it's not held during `next()`, and python locks are super cheap, so this definitely isn't worth it. I spent like 30 minutes reading through python memory consistency models to figure out if this code had a bug, and I think I landed on in cPython, it happens to work at the moment, but is not guaranteed.

See https://peps.python.org/pep-0583/#uninitialized-values-flag

You're relying on a separate variable (`self.finished`) to act as a flag for a variable set in another thread (`self.results`) which is only true in a globally sequentially consistent memory model without synchronization. AFAIK (that PEP didn't really have a conclusion that I could see), Python does not guarantee such a model, nor does cPython, though it happens to likely have it at the moment.

I think we should lock liberally and have known bug-free code, and focus on true contention (`next()`) vs micro-optimizations before we have a need.

Reply:
I'm glad you brought this up - I took another crack at it and the code is way simpler now - PTAL!